src/easylink/configuration.py

```diff
@@ -45,42 +85,54 @@ class Config:
 
     def __init__(
         self,
-        pipeline_specification: Union[str, Path],
-        input_data: Union[str, Path],
-        computing_environment: Optional[Union[str, Path]],
-        results_dir: Union[str, Path],
+        config_params: Dict,
     ):
-        # Handle pipeline specification
-        self.pipeline = load_yaml(pipeline_specification)
-        self._requires_spark = self._determine_if_spark_is_required(self.pipeline)
-        # Handle input data specification
-        self.input_data = self._load_input_data_paths(input_data)
-        # Handle environment specification
-        self.environment = self._load_computing_environment(computing_environment)
-        # Store path to results directory
-        self.results_dir = Path(results_dir)
-
-        # Extract environment attributes and assign defaults as necessary
-        self.computing_environment = self._get_required_attribute(
-            self.environment, "computing_environment"
-        )
-        self.container_engine = self._get_required_attribute(
-            self.environment, "container_engine"
-        )
-        self.slurm = self.environment.get("slurm", {})  # no defaults for slurm
-        self.implementation_resources = self._get_implementation_resource_requests(
-            self.environment
-        )
-        self.spark = self._get_spark_requests(self.environment, self._requires_spark)
+        super().__init__(layers=["initial_data", "default", "user_configured"])
+        self.update(DEFAULT_ENVIRONMENT, layer="default")
+        self.update(config_params, layer="user_configured")
+        for step in self.pipeline["steps"]:
```

Comment:
Where is `self.pipeline` defined? Is it a key in the config tree? If so, won't we throw an error here if the user-defined config doesn't have that key, when any errors due to user error should get thrown by the `_validate` function below?

Reply:
It is a key in the config tree. The "pipeline" key itself is set in the configparams:
```
    return {
        "pipeline": load_yaml(pipeline_specification),
        "input_data": load_input_data_paths(input_data),
        "environment": load_computing_environment(computing_environment),
        "results_dir": Path(results_dir),
    }
```
independently of the user yaml.
(ln 42) 
The pipeline specification itself is a required CLI param, so that's when an error would be thrown if a user doesn't pass in a spec file at all.

Reply:
but "steps" isn't guaranteed yet!

Reply:
@zmbc Do we need the "steps" outer key in the pipeline yaml? Are we going to have some other outer key at any point?

Reply:
It doesn't seem like this key is in the Milestones doc at all, looking at it

Reply:
Yeah I don't remember having an outer key!